### PR TITLE
ol + ol-ext upgrade

### PR DESCRIPTION
**Pull request by jansule**
This upgrades to ol@10 and ol-ext@4.

### CODE REVIEW

File: src/modules/featurepopup/featurepopup-directive.js [hunk 1]
```diff
@@ -214,8 +216,14 @@ angular.module('anol.featurepopup')
                                     return;
                                 }
 
-                                var _features = layer.olLayer.getSource()
-                                    .getFeaturesInExtent(extent)
+                                let _featuresInExtent;
+                                if (layer.olLayer.getSource() instanceof VectorTile) {
```
Comment (simonseyock) [inline src/modules/featurepopup/featurepopup-directive.js:220]:
```suggestion
                                if (layer.olLayer instanceof VectorTileLayer) {
```

File: src/modules/featurepopup/featurepopup-directive.js [hunk 2]
```diff
@@ -2,6 +2,7 @@ import './module.js';
 import '../util';
 import Overlay from 'ol/Overlay';
 import Cluster from 'ol/source/Cluster';
+import VectorTile from 'ol/source/VectorTile';
```
Comment (simonseyock) [inline src/modules/featurepopup/featurepopup-directive.js:5]:
```suggestion
import VectorTileLayer from 'ol/layer/VectorTile';
```

File: src/modules/featurepopup/featurepopup-directive.js [hunk 3]
```diff
@@ -214,8 +216,14 @@ angular.module('anol.featurepopup')
...
 
-                                var _features = layer.olLayer.getSource()
-                                    .getFeaturesInExtent(extent)
+                                let _featuresInExtent;
+                                if (layer.olLayer.getSource() instanceof VectorTile) {
+                                  _featuresInExtent = layer.olLayer.getFeaturesInExtent(extent);
+                                } else {
+                                  _featuresInExtent = layer.olLayer.getSource()
+                                    .getFeaturesInExtent(extent);
+                                }
```
Comment (simonseyock) [inline src/modules/featurepopup/featurepopup-directive.js:225]:
```suggestion
                                if (layer.olLayer instanceof VectorTileLayer) {
                                  _featuresInExtent = layer.olLayer.getFeaturesInExtent(extent);
                                } else if (layer.olLayer.getSource() instanceof VectorSource) {
                                  _featuresInExtent = layer.olLayer.getSource()
                                    .getFeaturesInExtent(extent);
                                }
```
I would prefer if the if would check for the class that actually contains `getFeaturesInExtent`, if that works.

The `else if` is probably not necessary.

**Review by simonseyock — APPROVED** [review]
very nice!

Only the `refresh` function issue really needs to be addressed in my opinion.

File: src/anol/layer.js [hunk 4]
```diff
@@ -167,9 +167,8 @@ class AnolBaseLayer {
         angular.element(this).off('anol.layer.visible:change', func);
     }
     refresh() {
-        if(this.olLayer instanceof BaseLayer && this.olLayer.getSource() instanceof Source) {
-            var source = this.olLayer.getSource();
-            source.refresh();
+        if(this.olLayer instanceof BaseLayer) {
+            this.olLayer.changed();
         }
     }
```
[the remark this answers is not in the document]
Comment (jansule) [reply src/anol/layer.js:173]: refresh is only implemented as re-rendering in the base class. The other classes implement the actual refresh logic.

File: src/modules/featurepopup/featurepopup-directive.js (hunk 3 above)
Comment (jansule) [reply src/modules/featurepopup/featurepopup-directive.js]: done

File: src/modules/featurepopup/featurepopup-directive.js (hunk 2 above)
Comment (jansule) [reply src/modules/featurepopup/featurepopup-directive.js]: done

**Pull request merged**